### PR TITLE
[FEAT] Expand code Link

### LINKED ISSUE #26: Expand Code Links
## Why do we need this ?

As a GitHub App developer

**I need to**: 
Add a listener for room messages and extend GitHub Code links like - https://github.com/RocketChat/Apps.Github22/blob/main/github/definitions/githubIssueData.ts#L4-L9 
**So That**:
Users can see the linked code segment ins the sent message.
## Acceptance Criteria

- [ ] Add listener for room messages which contain GitHub Code Segment link
- [ ] Extend the message to contain the linked code.

## Further Comments / References  

- [Documentation](https://rocketchat.github.io/Rocket.Chat.Apps-engine/modules/messages_ipremessagesentextend.html)
- Inspiration : 
image

### PR DESCRIPTION
**Pull request by VipinDevelops**
## Issue(s) 
Closes #26 

## Acceptance Criteria fulfillment

This pull request focuses on fulfilling the acceptance criteria related to expanding code links .

The changes introduced in this PR involve the following steps:

- The `checkPreMessageSentExtend` function first verifies whether the received message contains a GitHub link.
- The `executePreMessageSentExtend` function then checks if the identified link is a code link. 
- Finally, the` handleCodeLink` function is responsible for adding an attachment to the message, thereby expanding the code link.
## Proposed changes (including videos or screenshots)

image

[Screencast from 2023-06-22 11-46-22.webm](https://github.com/RocketChat/Apps.Github22/assets/99081689/4192a380-8e8b-4d0e-9f53-64f5a59344ee)

### CODE REVIEW

**Review by samad-yar-khan — CHANGES_REQUESTED** [review]
@VipinDevelops  Thanks for the PR. This is an amazing contribution. Please go through the requested changes.

File: github/GithubApp.ts [hunk 1]
```diff
@@ -45,12 +46,41 @@ import { IJobContext } from "@rocket.chat/apps-engine/definition/scheduler";
 import { IRoom } from "@rocket.chat/apps-engine/definition/rooms";
 import { clearInteractionRoomData, getInteractionRoomData } from "./persistance/roomInteraction";
 import { GHCommand } from "./commands/GhCommand";
+import { IPreMessageSentExtend, IMessage } from "@rocket.chat/apps-engine/definition/messages";
+import { handleCodeLink } from "./handlers/HandleLinks";
+import { isGithubLink, hasCodeLink } from "./helpers/checkLinks";
 
-export class GithubApp extends App {
+export class GithubApp extends App implements IPreMessageSentExtend {
```
[the remark this answers is not in the document]
Comment (VipinDevelops) [reply github/GithubApp.ts:53]: Yeah I tried searching for something else but implementing the `IPreMessageSentExtend `interface in the `GithubApp` class is the remains approach,  We do not have any other way  for PreMessageSent

File: github/GithubApp.ts [hunk 2]
```diff
@@ -45,12 +46,41 @@ import { IJobContext } from "@rocket.chat/apps-engine/definition/scheduler";
...
+        message: IMessage,
+        extend: IMessageExtender,
+        read: IRead,
+        http: IHttp,
+        persistence: IPersistence
+    ): Promise<IMessage> {
+
+        if (await hasCodeLink(message)) {
+            await handleCodeLink(message, read, http, message.sender, message.room, extend);
+        }
```
[the remark this answers is not in the document]
Comment (VipinDevelops) [reply github/GithubApp.ts:79]: Sure Ill update the name's

File: github/GithubApp.ts [hunk 3]
```diff
@@ -68,10 +98,9 @@ export class GithubApp extends App {
             },
         };
         let text = `GitHub Authentication Succesfull 🚀`;
-        let interactionData = await getInteractionRoomData(
-            read.getPersistenceReader(),
-            user.id
-        );
+
+        let interactionData = await getInteractionRoomData(read.getPersistenceReader(), user.id);
+
```
[the remark this answers is not in the document]
Comment (VipinDevelops) [reply github/GithubApp.ts]: sure I Fixed This

File: github/handlers/HandleLinks.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    user: IUser,
+    room: IRoom,
+    extend: IMessageExtender
+) {
+    const regex: RegExp = /\bhttps?:\/\/github\.com\/\S+\b/;
+    let text = message.text!;
+    const match: RegExpMatchArray | null = text.match(regex);
+    const result: string | undefined = match?.[0];
+    let url: string = result?.replace("blob/", "")!;
+    url = url.replace("github.com", "raw.githubusercontent.com");
```
[the remark this answers is not in the document]
Comment (VipinDevelops) [reply github/handlers/HandleLinks.ts]: Sure I made enums for this

File: github/handlers/HandleLinks.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        );
+        const text = await content.match(lines);
+        return text[1];
+    } else {
+        const lines = new RegExp(`(?:.*\n){0}(.*(?:\n.*){5})`);
+        const text = await content.match(lines);
+        return text[1];
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (VipinDevelops) [reply github/handlers/HandleLinks.ts]: So I should introduce and implement the Test for this using JEST or something ?

**Comment by VipinDevelops** [comment]
Hello @samad-yar-khan  Sir i made all the requested changes in the PR

File: github/handlers/GitHubCodeSegmentHandler.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import { GitHubURLEnum } from "../enum/GitHubURL";
+
+async function extractCodeSnippetFromURL(content: string, url: string): Promise<string> {
+    const lineRangeRegex: RegExp = /(?:L(\d+)+-L(\d+)|L(\d+))/;
+    const lineRangeMatch: RegExpMatchArray | null = url.match(lineRangeRegex);
+
+    if (lineRangeMatch) {
+    return extractCodeSnippetByLineRange(content, lineRangeMatch);
+    }
+
```
Comment (samad-yar-khan) [inline github/handlers/GitHubCodeSegmentHandler.ts:16]: Missing some indentation
```
if (lineRangeMatch) {
    return extractCodeSnippetByLineRange(content, lineRangeMatch);
}
```

**Review by samad-yar-khan — CHANGES_REQUESTED** [review]

**Review by samad-yar-khan — COMMENTED** [review]
This is an amazing contribution to our project ! Thanks a lot @VipinDevelops

**Review by samad-yar-khan — APPROVED** [review]
LGTM !

**Pull request merged**